packages/block-directory/package.json

```diff
@@ -21,10 +21,6 @@
 	"main": "build/index.js",
 	"module": "build-module/index.js",
 	"react-native": "src/index",
-	"sideEffects": [
```

Comment:
This package confuses me. How is it supposed to be used?

Is it meant to be a single side effect, by doing just `import '@wordpress/block-directory';`? I'm seeing a bunch of components in there that don't seem to be loaded directly, but I guess they're all loaded transitively?

Reply:
`block-directory` is a JS plugin. You just load the script and it hooks into the `wp.data`, `wp.hooks` and `wp.plugins` APIs, adding a new feature to the editor.

If you were building your own app with the NPM packages, the package would be activated by a simple side-effectful import, as you wrote. The APIs it hooks into are (ideally) peer dependencies of the `block-editor` package.

Reply:
> I'm seeing a bunch of components in there that don't seem to be loaded directly, but I guess they're all loaded transitively?

These are probably all used to implement the plugin UI, rather than being exported for 3rd party consumers. Are there any specific components that look suspicious?

Reply:
None in particular. I expect them all to be used to implement the plugin UI, yes, unless one of them got dropped over time.

Reply:
We could make it side effects free. I didn't consider it when refactoring to behave like a plugin. Edit family of packages uses `initialize` methods and we could offer something similar. Definitely a follow up task if you feel like it's worth the hassle. 

Would it be useful if more functionality would be initialized this way?

Reply:
Great question, @gziolo! 

In general, I feel that any package that could be classified as a library/plugin/utility (that is, something that doesn't invert control) should avoid side effects as much as possible.

Even setting aside the issues with bundlers for a second, initialisation functions generally provide more flexibility, because the consumer has more control over when they're invoked. Consumers can e.g. import something statically and call `initialize()` only when the editor is being loaded; without an initialisation method, they'd have to instead use a dynamic import, which by its async nature could be hard to fit into an existing codebase.

In terms of difficulty, it's generally a trivial amount of extra work for the naive use-case anyway. It's really not too bad to go from:

```js
import 'side-effectful-module';
```

to

```js
import { initialize } from 'side-effect-free-module';
initialize();
```

This way, users who don't care still get their initialisation at the top level of their module, while users who do care can move it elsewhere.

So while I don't think it's anything urgent that needs immediate attention, I'd definitely recommend switching to initialisation methods whenever the next code revamp opportunity comes along 🙂

Reply:
> Would it be useful if more functionality would be initialized this way?

What @sgomes said 🙂 I'll only add that the inflexible initialization was sometimes causing problems in the first Gutenlypso integration 2 years ago. The problematic code usually looks like this:
```js
import '@wordpress/rigid';
import flexible from '@wordpress/flexible';

getInitData().then( initData => {
  flexible.init( initData );
  // I'd love to do the following, but can't
  // rigid.init();
} );
```

The `rigid` package needs to be initialized only after `flexible` (because, for example, it does `select( 'core/flexible' )` during initialization), but I can't do that. `import` statements must be at the top level of the module, not inside a callback.

Solutions are either using `require( '@wordpress/rigid' )`, or creating an extra async-loaded module that is loaded just-in-time.

On the other hand, I think that declaring `sideEffects` on absolutely everything is not needed. `sideEffects: false` is useful for libraries where you can use just a little part of the library and want to avoid bundling the rest.

For plugins like `block-directory` or apps like `edit-post` we can keep the default where everything is assumed to have side effects, and I don't think it causes much deoptimization if any.

Reply:
In WordPress core we would have to call `wp.blockLibrary.initialize()` but it wouldn't be that much work. It's definitely something to keep in mind when working on new packages.